### ISSUE TITLE
Strategies to preallocate pages for real-time applications?

### ISSUE THREAD

**Issue by glenn-elliott** [jemalloc/jemalloc#2442]
I am developing a real-time application that uses jemalloc and runs on the PREEMPT_RT Linux kernel.  A best-practices document from the PREEMPT_RT project suggests that an application should [preallocate its heap](https://archive.kernel.org/oldwiki/rt.wiki.kernel.org/index.php/Dynamic_memory_allocation_example.html) before starting real-time work.  The provided example is for a glibc-managed heap.

I want to achieve something similar for my jemalloc-using application, with the goal of obtaining all of the pages from the OS that I expect jemalloc will need before the application starts real-time work.  Does this question make sense in context of jemalloc's arena-based design?  Unless I have overlooked something, I don't see any public API for merely allocating pages.  If I disable `dirty_decay_ms`, can my problem be solved simply by allocating an enormous chunk of memory with jemalloc, fault every page, and free it?  I'm not expecting to be so lucky.

It is feasible for me to generate a profile of my applications's heap needs with calls to `mallctl()`, though I don't think it would be fruitful for me to attempt to precisely describe the total number and size of all allocations in my application.  I cannot guarantee the same precise allocation pattern for my application in every instance would run, though I can estimate a high water mark for dynamically allocated memory.

A [question](https://github.com/jemalloc/jemalloc/issues/1375) about preallocation has come up before, but in the context of persistent memory.  Moreover, the author of that question wanted to reconstruct the jemalloc allocation state.  I merely seek to obtain pages from the OS ahead of time.

**Comment by daniel-brosche** [comment]
I always wanted to use jemalloc for realtime applications. Is there a way to create a private arenas where the max size is limited and preallocated / reserved for more deterministic allocations?

**Comment by glenn-elliott** [comment]
Here's what I ended up having to do:
1. Implement an extents allocator plugin. This plugin allocates memory with `mmap(MAP_POPULATE)`, and then doles it out as extents are allocated.
2. Patch jemalloc extents allocator API to tell me how big of an object it's trying to allocator. Deny this request if the total number of active pages is significantly less than the number of pages already allocated by extents. My understanding is that jemalloc assumes overcommit, so it is agressive in its extents requests.

(I tried disabling freeing of pages to the OS by setting `dirty_decay_ms` and `muzzy_decay_ms` to `-1`, but I ran into a strange edge case where an allocation freed by a thread that was not the one did the allocation would result in an effective memory leak. My alloc-aggressive programs OOM'ed within minutes.)
